File: multiplePointersPattern/countUniquesValues/unique.rb
def count_unique_values(ary)
    return ary.length if ary.length < 2

    i = 0
    unique_list = [ary[0]]
    for j in (1...ary.length)
        if unique_list[i] != ary[j]
            unique_list.push(ary[j])
            i += 1
        end
    end
    return i + 1
end


puts(count_unique_values([1, 1, 1, 1, 1, 1, 2])) # 2
puts(count_unique_values([1, 2, 3, 4, 4, 4, 7, 7, 12, 12, 13])) # 7
puts(count_unique_values([])) # 0
puts(count_unique_values([-2, -1, -1, 0, 1])) # 4
puts(count_unique_values([1])) # 1

File: multiplePointersPattern/countUniquesValues/unique.js
function countUniqueValues(ary) {
  if (ary.length < 2) return ary.length;

  let i = 0;
  uniqueAry = [ary[0]];
  for (let j = 1; j < ary.length; j++) {
    if (uniqueAry[i] !== ary[j]) {
      uniqueAry.push(ary[j]);
      i++;
    }
  }
  return i + 1;
}

countUniqueValues([1, 1, 1, 1, 1, 1, 2]); // 2
countUniqueValues([1, 2, 3, 4, 4, 4, 7, 7, 12, 12, 13]); // 7
countUniqueValues([]); // 0
countUniqueValues([-2, -1, -1, 0, 1]); // 4
